judoscale/django/middleware.py

```diff
@@ -12,10 +12,9 @@ def __init__(self, get_response):
     def __call__(self, request):
         request_start_header = request.META.get("HTTP_X_REQUEST_START", "")
 
-        request_metric = RequestMetrics(request_start_header)
-        metric = request_metric.get_queue_time_metric_from_header()
-        if metric:
-            metrics_store.add(metric)
+        if metric := Metric.for_web(request_start_header):
+            if collector := reporter.find_collector(WebMetricsCollector):
```

Comment:
It feels wrong to me that the middleware gets the collector from the reporter. I'm also concerned about the performance impact of running `find_collector` with every web request (which might be negligible).

Is there a way to pass the collector to the middleware's initializer, similar to how you [set it up for Flask](https://github.com/judoscale/judoscale-python/pull/29/files#diff-063a0f790822be3ab320856bb6c29bc1ebfbda52c7588a350d5005b63823d1e7R30)?

If not, is there an idiomatic way to have a singleton instance of `WebMetricsCollector`, since we'll only ever have one?

Reply:
> It feels wrong to me that the middleware gets the collector from the reporter.
> Is there a way to pass the collector to the middleware's initializer, similar to how you set it up for Flask?

I'll have a deeper look and report back.

> I'm also concerned about the performance impact of running find_collector with every web request (which might be negligible).

I benchmarked this with Python's `timeit` module:
```
With a single collector, uncached:                                 92 nsec average
With four collectors, WebMetricsCollector first item, uncached:    94 nsec average
With four collectors, WebMetricsCollector last item, uncached:    460 nsec average
With 100 collectors, WebMetricsCollector first item, uncached:     92 nsec average
With 100 collectors, WebMetricsCollector last item, uncached:   12000 nsec average
Any variation from above, cached with @functools.cache:            55 nsec average
```

So one option is to cache the collector lookups. But I'm leaning towards seeing if there is better way, either via a singleton instance or some other mechanism.

Reply:
Turns out I had misunderstood how the Django middleware gets instantiated — it's done only once per process and each HTTP request is "handled" by the `__call__` method. This means we instantiate and store the collector in the middleware initialiser, as well as register it with the reporter, as in this commit: https://github.com/judoscale/judoscale-python/pull/29/commits/f46f6b5b0b62b0a064d7ab1c94924bd1de28cd13.

That looks like a better solution to me than the `.find_collector()` route.

Reply:
Yep, I think that's great! I also assumed the middleware was instantiated for each request, since that's how Rack middleware works in the Ruby world.

Is there a reason to keep `find_collector` around now?